File: pharmacy-api/local-api/middleware/middleware.go
package middleware

import (
	"errors"
	"log"
	"net/http"
	"os"
	local_models "pharmacy/local-api/models"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/golang-jwt/jwt/v4"
	"gorm.io/gorm"
)

var (
	jwtSecret = []byte(os.Getenv("ACCESSTOKEN_SECRET"))
)

func AuthMiddleware(db *gorm.DB, requiredPermission string) gin.HandlerFunc {
	return func(c *gin.Context) {
		authHeader := c.GetHeader("Authorization")
		if authHeader == "" {
			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "Authorization header required"})
			return
		}

		parts := strings.Split(authHeader, " ")
		if len(parts) != 2 || parts[0] != "Bearer" {
			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "Invalid authorization format"})
			return
		}
		tokenString := parts[1]

		claims := &local_models.Claims{}
		token, err := jwt.ParseWithClaims(tokenString, claims, func(token *jwt.Token) (any, error) {
			return jwtSecret, nil
		})

		if err != nil || !token.Valid {
			if err == jwt.ErrTokenExpired {
				c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "token_expired"})
				return
			} else {
				c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "Invalid token"})
				return
			}
		}
		if requiredPermission != "" {
			var role local_models.Role
			if err := db.Preload("Permissions").
				Where("roles.name = ?", claims.Role).Find(&role).Error; err != nil {
				if errors.Is(err, gorm.ErrRecordNotFound) {
					log.Println("Middleware error [" + claims.Username + "]" + err.Error())
					c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "Role not found"})
					return
				}
				log.Println("Middleware error [" + claims.Username + "]" + err.Error())
				c.AbortWithStatusJSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
				return
			}
			for _, permission := range role.Permissions {
				if permission.Action == requiredPermission {
					c.Set("user", claims)
					c.Next()
					return
				}
			}

			c.AbortWithStatusJSON(http.StatusForbidden, gin.H{"error": "Permission denied"})
		} else {
			c.Set("user", claims)
			c.Next()
			return
		}
	}
}

File: pharmacy-api/src/local-api/middleware/middleware.ts
import { Request, Response, NextFunction } from 'express';
import { DataSource } from 'typeorm';
import * as jwt from 'jsonwebtoken';
import { Claims } from '@pharmacy/src/shared/models/models';
import { Role, Permission } from '@pharmacy/src/shared/models/database_models';

const jwtSecret = process.env.ACCESSTOKEN_SECRET;

declare global {
  namespace Express {
    interface Request {
      user?: Claims;
    }
  }
}

export function authMiddleware(dataSource: DataSource, requiredPermission: string = '') {
  return async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    const authHeader = req.headers['authorization'];
    
    if (!authHeader) {
      res.status(401).json({ error: "Authorization header required" });
      return;
    }

    const parts = authHeader.split(' ');
    if (parts.length !== 2 || parts[0] !== 'Bearer') {
      res.status(401).json({ error: "Invalid authorization format" });
      return;
    }

    const tokenString = parts[1];

    try {
      const decoded = jwt.verify(tokenString, jwtSecret) as Claims;
      const claims = decoded;

      if (!requiredPermission) {
        req.user = claims;
        next();
        return;
      }

      const roleRepository = dataSource.getRepository(Role);
      const role = await roleRepository.findOne({
        where: { name: claims.role },
        relations: ['permissions'],
      });

      if (!role) {
        console.log(`Middleware error [${claims.username}] Role not found: ${claims.role}`);
        res.status(401).json({ error: "Role not found" });
        return;
      }

      const hasPermission = role.permissions?.some(
        (permission: Permission) => permission.action === requiredPermission
      );

      if (!hasPermission) {
        console.log(`Middleware error [${claims.username}] Permission denied: ${requiredPermission}`);
        res.status(403).json({ error: "Permission denied" });
        return;
      }

      req.user = claims;
      next();
      
    } catch (err) {
      if (err instanceof jwt.TokenExpiredError) {
        res.status(401).json({ error: "token_expired" });
      } else if (err instanceof jwt.JsonWebTokenError) {
        res.status(401).json({ error: "Invalid token" });
      } else {
        console.error(`Middleware error:`, err);
        res.status(500).json({ error: "Internal server error" });
      }
      return;
    }
  };
}